### DISCUSSION TITLE
Bad performance with model binding versus computed properties

### DISCUSSION THREAD

**Discussion by ryatkins** [livewire/livewire#3484] [1. Help]
I have a large data set that I'm eager loading with a computed property and this works well from a performance standpoint.  Mounting and updates on the page happen quickly as well.

I would prefer to use model binding so that it is easier to update some of these rows as I can wire:model directly to them.

When I try to bind model properties the update performance drops off a cliff.   The hydration portion of the lifecycle takes more than 30 seconds and my dev server execution time maxes out.  I'm not even updating data yet, I'm just trying to validate the input on another field, but my large dataset is hydrating first.

Is there a best of both worlds that I can get better performance out of model binding? 

Thanks!

**Comment by Dimonka2** [comment]
You should give user opportunity to navigate using pagination/sorting/filtering and not load whole internet into single page. This step will probably solve your performance issue. 

Regarding the idea of mounting models - I am not a big fan of it and still think that you have to load just what you need to show. Models having usually a lot of other fields that you even do not need to load from DB to show records. Otherwise you passing whole data twice - in formatted way and in Livewire "serverMemo" and if you have a DebugBar it gets replicated there as well... And then every request all this data is going to be passed back-and-forth. Thus.. If you still really want to make an edit for those models, than it makes sense to break down your page structure into two components: parent - list of items, child - your model element. Then by changing one model you will reload only one model related html and data.

This is how I see it.

**Comment by ryatkins** [comment]
I've narrowed this down to a nested rule that is timing out and I can't figure out why. 

`'boms.*.qty_per_par' => 'required`

```
  public function updatedBoms()
    {
        $this->validate();
        // $this->validateOnly('boms.0.qty_per_par');  // also fails
    }
```

**Comment by joshhanley** [comment]
@ryatkins I've got no idea what could be causing that! It will probably take some debugging with xdebug to try and work out where exactly the code is hanging.

Is it hanging when using the new deep nested model stuff? If so, can you try it on the latest version of Livewire instead and see if it still happens?

**Comment by ryatkins** [comment]
Hi Josh - it’s happening on both the latest release and also on dev-master
with the nested code.

I’ll keep digging in!

**Reply by joshhanley** [reply]
@ryatkins ah ok, good in a way? Haha at least I hadn't broken something, it was already happening 🙈

Is that the full rule or is there more checks, as it's missing a `'` off the end?

**Reply by ryatkins** [reply]
Well, it’s consistent on both branches  :)

I did have more to the rule, integer and gt:0 but it is happening regardless.  

Where do you recommend loading the model in the lifecycle, only on mount?

**Reply by joshhanley** [reply]
Hmm it's odd. Can you dump what `boms.*.pty_per_par` data looks like before the validation is run so we can check it out?

Yeah mount should be fine.

**Comment by ryatkins** [comment]
So I found the issue and I've run into it before, but it continues to bite me.  

Livewire (and perhaps Laravel) doesn't preserve the original relationship parameters during the re-hydrate.   

Clauses that were originally set on mount are not included in the hydrate: 
 - the limited select call 
 - `->where('imrefdes_parent_item', $this->item_no)` is missing.  This was resulting in 16k rows being called.    

This doesn't happen with a computed property because my explicit query is called every time, but it means I can't easily bind to model properties.   

```
       $boms = bmprdstr::
         ->with([
           'references' => function ($query) {
                $query->select('ID', 'imrefdes_note', 'imrefdes_item_no')
                    ->where('imrefdes_parent_item', $this->item_no);
            },
        ])
        ->where('item_no', $this->item_no)->get();
```

**Reply by joshhanley** [reply]
@ryatkins ah! Right yep, Livewire has no way to know when custom queries have been run, as all it can see is the result of the query.

The information Livewire tracks on the front end, is the class and ID's of any models in public properties. If relations are also loaded, it tracks which relations have been loaded, but not specific relation models.

Livewire uses Laravel's queue serialisation for this, so if you were to put a job onto the queue with the same query you'd probably experience the same issue.

So I had wrote out the below, as it is a way to use custom hydration for properties, but I had a quick look at the code, and I think this would actually wipe out the dirty data from any of the model binding updates, so I don't actually think it will work.

Might be worth experimenting though.

If it does wipe it out, maybe it's something that can be explored, which is running the custom hydration method before dirty data gets applied.

Hope this helps!

---

~~But there is a way around it, you could use a custom hydration for that collection.~~

Imagine you had a `$users` property which had some custom query to get them all with some joins or something, I would structure the component like this.

```php
public $users;

public function mount() {
    $this->getUsers();
}

public function hydrateUsers() {
    $this->getUsers();
}

protected function getUsers() {
    $this->users = User::all(); // Imagine this was some complex query
}
```
